pkg/reconciliation/reconcile_racks_test.go

```diff
@@ -2952,3 +2952,99 @@ func TestCheckRackLabels(t *testing.T) {
 	require.Equal(result.Continue(), res, "Label updates should not cause errors")
 	require.Subset(desiredStatefulSet.Labels, rc.Datacenter.GetRackLabels("default"))
 }
+
+func TestCheckPodsReadyAllStarted(t *testing.T) {
+	rc, _, cleanupMockScr := setupTest()
+	defer cleanupMockScr()
+	assert := assert.New(t)
+
+	desiredStatefulSet, err := newStatefulSetForCassandraDatacenter(
+		nil,
+		"default",
+		rc.Datacenter,
+		3)
+	assert.NoErrorf(err, "error occurred creating statefulset")
+
+	desiredStatefulSet.Status.ReadyReplicas = *desiredStatefulSet.Spec.Replicas
+
+	trackObjects := []runtime.Object{
+		desiredStatefulSet,
+		rc.Datacenter,
+	}
+
+	mockPods := mockReadyPodsForStatefulSet(desiredStatefulSet, rc.Datacenter.Spec.ClusterName, rc.Datacenter.Name)
+	for idx := range mockPods {
+		mp := mockPods[idx]
+		metav1.SetMetaDataLabel(&mp.ObjectMeta, api.SeedNodeLabel, "true")
+		trackObjects = append(trackObjects, mp)
+	}
+
+	rc.Client = fake.NewClientBuilder().WithStatusSubresource(rc.Datacenter).WithRuntimeObjects(trackObjects...).Build()
+
+	nextRack := &RackInformation{}
+	nextRack.RackName = desiredStatefulSet.Labels[api.RackLabel]
+	nextRack.NodeCount = int(*desiredStatefulSet.Spec.Replicas)
+	nextRack.SeedCount = 1
+
+	rackInfo := []*RackInformation{nextRack}
+
+	rc.desiredRackInformation = rackInfo
+	rc.statefulSets = make([]*appsv1.StatefulSet, len(rackInfo))
+	rc.statefulSets[0] = desiredStatefulSet
+
+	rc.clusterPods = mockPods
+	rc.dcPods = mockPods
+
+	epData := httphelper.CassMetadataEndpoints{
+		Entity: []httphelper.EndpointState{},
+	}
+
+	for i := 0; i < int(*desiredStatefulSet.Spec.Replicas); i++ {
+		ep := httphelper.EndpointState{
+			RpcAddress: fmt.Sprintf("192.168.1.%d", i+1),
+			Status:     "UN",
+		}
+		epData.Entity = append(epData.Entity, ep)
+	}
+
+	res := &http.Response{
```

Comment:
Suggestion: would it be possible to separate out the code which fakes the reconciliation from the code which asserts the intended state? I'm struggling to figure out what the intention of some of this stuff is to be honest.

Reply:
Considering I'm assuming that the reconciliation-faking code is common across all of these tests, it might even be nice to run it in a separate controller in a goroutine, so that we can factor it out of all of these tests in the longer term.

Reply:
For separate controller test we have envtests, they wouldn't use fake client. And no, the states are not common between all tests (as these only test a single check inside the controller flow). Think of these as unit tests.

Reply:
So each test has its own code with only the parts set that are actually required in the method that's being tested. Any other extra should be removed to  catch the modification of the implementation.

Reply:
OK, I sort of understand your point here. I'm unclear on what this is testing though - `TestCheckPodsReadyAllStarted` isn't quite describing what additional scenario you're testing here. Can you explain what additional testing you're doing and why this PR needs it?

Reply:
There was no existing test for CheckPodsReady() at all in these tests. So I wanted a test that ensures the behavior does not change in the normal scenario where everything is already up (not by this or any other PR by accident). 

It's a safeguard and documentation what's required for that method to work. If you start requiring some additional data in that method (such as by adding new function call in that method, which itself is already quite horrible code) then the test will fail. 

It does not test any new behavior, there simply wasn't anything to test its current requirements. This PR does not need it for other than ensuring I didn't break it when I reversed the logic.

Reply:
OK, sounds fine then.